WordPressAuthenticator/OAuth/ProofKeyForCodeExchange.swift

```diff
@@ -25,21 +25,92 @@ struct ProofKeyForCodeExchange {
-            fatalError()
+            // The spec defines code_challenge for the s256 mode as:
+            //
+            // code_challenge = BASE64URL-ENCODE(SHA256(ASCII(code_verifier)))
+            //
+            // We don't need the ASCII conversion, because we build `CodeVerifier` from URL safe
+            // characters.
+            let rawData = Data(codeVerifier.rawValue.utf8)
+            let hashedData: Data = rawData.sha256Hashed()
+            return hashedData.base64URLEncodedString()
         case .plain:
-            return codeVerifier
+            return codeVerifier.rawValue
+        }
+    }
+}
+
+extension ProofKeyForCodeExchange {
+
+    struct CodeVerifier: Equatable {
+
+        let rawValue: String
+
+        static let allowedCharacters = Character.urlSafeCharacters
+        static let allowedLengthRange = (43...128)
+
+        /// Generates a random code verifier according to the PKCE RFC.
+        ///
+        /// - Note: This method name is more verbose than the recommended "make<Type>" for this factory to communicate the randomness component.
+        static func makeRandomCodeVerifier() throws -> Self {
+            let value = try String.randomSecureCodeVerifier()
+
+            // It's appropriate to force unwrap here because a `nil` value could only result from
+            // a developer error—either wrong coding of the constrained length or of the allowed
+            // characters.
+            return .init(value: value)!
         }
+
+        init?(value: String) {
+            guard CodeVerifier.allowedLengthRange.contains(value.count) else { return nil }
+
+            guard Set(value).isSubset(of: CodeVerifier.allowedCharacters) else { return nil }
+
+            self.rawValue = value
+        }
+    }
+}
+
+private extension String {
+
+    /// Generates a random code verifier according to the PKCE RFC.
+    ///
+    /// The RFC states:
+    ///
+    /// > It is RECOMMENDED that the output of a suitable random number generator be used to create a 32-octet sequence.
+    /// > The octet sequence is then base64url-encoded to produce a 43-octet URL safe string to use as the code verifier.
+    static func randomSecureCodeVerifier() throws -> String {
+        let byteCount = 32
+        var bytes = [UInt8](repeating: 0, count: byteCount)
+        let result = SecRandomCopyBytes(kSecRandomDefault, byteCount, &bytes)
+
+        guard result == errSecSuccess else {
+            throw OAuthError.failedToGenerateSecureRandomCodeVerifier(status: result)
+        }
+
+        let data = Data(bytes)
+
+        // Base64url-encoding a 32-octect sequence should always result in a 43-length string,
+        // string, but let's cap it just in case.
+        //
+        // Also notice that by base64url-encoding, we ensure the characters are in the allowed
+        // set.
+        return String(data.base64URLEncodedString().prefix(43))
```

Comment:
I thought about using the lower bound, but I though it would be confusing to use in relation to the "32-octed sequence base64url-encoded = 43 characters" assertion.

But I guess one could updated the comment a bit to explain that.

Will take a shot.

Reply:
330bed67